src/middleware/r2Middleware.ts

```diff
@@ -0,0 +1,153 @@
+import type { Context } from '../context';
+import { R2Provider } from '../providers/r2Provider';
+import responses from '../responses';
+import { hasTrailingSlash, isDirectoryPath } from '../utils/path';
+import type { Middleware } from './middleware';
+import latestVersions from '../constants/latestVersions.json' assert { type: 'json' };
+import type { Request } from '../routes/request';
+import { renderDirectoryListing } from '../utils/directoryListing';
+import { parseConditionalHeaders } from '../utils/request';
+import { once } from '../utils/memo';
+
+const getProvider = once((ctx: Context) => new R2Provider({ ctx }));
+
+export class R2Middleware implements Middleware {
+  async handle(request: Request, ctx: Context): Promise<Response> {
+    const path = getR2Path(request);
+    const isPathADirectory = isDirectoryPath(path);
+
+    return isPathADirectory
+      ? handleDirectory(request, path, ctx)
+      : handleFile(request, path, ctx);
+  }
+}
+
+async function handleDirectory(
+  request: Request,
+  r2Path: string,
+  ctx: Context
+): Promise<Response> {
+  if (!hasTrailingSlash(request.urlObj.pathname)) {
+    // We always want directory listing requests to have a trailing slash
+    const url = request.unsubtitutedUrl ?? request.urlObj;
+    return Response.redirect(url + '/', 301);
+  }
+
+  const result = await getProvider(ctx).readDirectory(r2Path, {
+    // /docs lists the nodejs/release directory - don't want to include the
+    //  files in there for that path
+    listFiles: !request.urlObj.pathname.startsWith('/docs'),
+  });
+
+  if (result === undefined) {
+    return responses.directoryNotFound(request);
+  }
+
+  if (result.hasIndexHtmlFile) {
+    // Prioritize showing index files over directory listings
+    return handleFile(request, r2Path + 'index.html', ctx);
+  }
+
+  let lastModified;
+  for (const file of result.files) {
+    if (lastModified === undefined || file.lastModified > lastModified) {
+      lastModified = file.lastModified;
+    }
+  }
+
+  let responseBody;
+  if (request.method === 'GET') {
+    responseBody = renderDirectoryListing(
+      request.unsubtitutedUrl ?? request.urlObj,
+      result
+    );
+  }
+
+  return new Response(responseBody, {
+    headers: {
+      'last-modified': (lastModified ?? new Date()).toUTCString(),
+      'content-type': 'text/html',
+      'cache-control': CACHE_HEADERS.success,
+    },
+  });
+}
+
+function handleFile(
+  request: Request,
+  r2Path: string,
+  ctx: Context
+): Promise<Response> {
+  switch (request.method) {
+    case 'HEAD':
+      return headFile(request, r2Path, ctx);
+    case 'GET':
+      return getFile(request, r2Path, ctx);
+  }
+
+  throw new Error('R2Middleware handleFile unsupported method');
+}
+
+async function headFile(
+  request: Request,
+  r2Path: string,
+  ctx: Context
+): Promise<Response> {
+  const result = await getProvider(ctx).headFile(r2Path);
+  if (result === undefined) {
```

Comment:
Sometimes, you add empty lines; sometimes, you do not (see this block and block below) when separating definition and comparison. Can we keep the empty lines consistent within the code? It is a minor nit but good for code readability.